runtime/src/iree/task/pool.c

```diff
@@ -85,23 +85,16 @@ static iree_status_t iree_task_pool_grow(iree_task_pool_t* pool,
   // zero-fill-on-demand trickery going on the pages are all wired here vs.
   // when the tasks are first acquired from the list where it'd be harder to
   // track.
-  uintptr_t p = ((uintptr_t)allocation + aligned_block_size) - pool->task_size;
-  iree_task_t* head = (iree_task_t*)p;
-  iree_task_t* tail = head;
-  head->next_task = NULL;
-  head->pool = pool;
-
-  // Work around a loop vectorizer bug that causes memory corruption in this
-  // loop. Only Android NDK r25 is known to be affected. See
-  // https://github.com/openxla/iree/issues/9953 for details.
-#if defined(__NDK_MAJOR__) && __NDK_MAJOR__ == 25
-#pragma clang loop unroll(disable) vectorize(disable)
-#endif
```

Comment:
@per Are you sure this is not needed anymore? When I looked at the generated loop code, it seemed to me that the heap corruption happened inside this loop, and I don't see how this patch modifies the loop body or the loop condition.

Reply:
I believe this to be a strict-aliasing rule violation. Casting the iree_task_allocation_header_t* buffer to a task_t is not allowed, but accessing the buffer allocated through a char* is ok.
Disabling strict-aliasing (which gets enabled with -O2) also makes the original code work, which indicates that the problem is strict-aliasing.
So I think this has nothing to do with the loop unrolling or vectorizer passes, but rather is a case of undefined behavior due to strict aliasing rules.

Reply:
Ah, I see now. I assumed that it was the other changes that made the difference.

I recreated my original android setup and checked the following:
- It passes with the code as in the `main` branch
- It still fails without the pragma
- It passes with `p` defined as `char *`, like in this PR

This is excellent, wish we had a strict aliasing sanitizer to check for issues like this.

I think that while `char *` solves the problem for us here, it's not sufficient according to C standard. This is because what really matters is the type used to access memory, and a proper fix would be to calculate the starting pointer with either `char *` or `uintptr_t` and then use `memcpy` to read/write those tasks. I mean something like:
```c
  char* p = ((char *)allocation + aligned_block_size) - pool->task_size;
  iree_task_t* tail = (iree_task_t *) p;
  iree_task_t* head = NULL;
  iree_task_t task = {};

  for (iree_host_size_t i = 0; i < actual_capacity; ++i) {
    memcpy(&task, p, sizeof(task));
    task.pool = pool;
    task.next_task = head;
    memcpy(p, &task, sizeof(task));
    head = (iree_task_t *) p;
    p -= pool->task_size;
  }
```

This also works for me.

Reply:
Agreed, TYSAN will hopefully catch these kind of things when it gets merged.

Reply:
@per Can you update the change to use memcpy (in the loop and the condition just after)? The patch looks good to me otherwise.

Reply:
In C you can reinterpret pointers as you wish as long as the effective type matches. 'Strict aliasing' is defined in terms of 'effective types', where what matters is type of memory accesses: C11 sec. 6.5 6-7:

```
6  The effective type of an object for an access to its stored value is the declared type of the
object, if any.87) If a value is stored into an object having no declared type through an
lvalue having a type that is not a character type, then the type of the lvalue becomes the
effective type of the object for that access and for subsequent accesses that do not modify
the stored value. If a value is copied into an object having no declared type using
memcpy or memmove, or is copied as an array of character type, then the effective type
of the modified object for that access and for subsequent accesses that do not modify the
value is the effective type of the object from which the value is copied, if it has one. For
all other accesses to an object having no declared type, the effective type of the object is
simply the type of the lvalue used for the access.

7 An object shall have its stored value accessed only by an lvalue expression that has one of
the following types:88)
— a type compatible with the effective type of the object,
— a qualified version of a type compatible with the effective type of the object,
— a type that is the signed or unsigned type corresponding to the effective type of the
object,
— a type that is the signed or unsigned type corresponding to a qualified version of the
effective type of the object,
— an aggregate or union type that includes one of the aforementioned types among its
members (including, recursively, a member of a subaggregate or contained union), or
— a character type.
````

My reading of this has always been that `char`/`memcpy` can be used to *access* (read/write) any memory object, otherwise the type read has to match the last type written. If this is the case:
> Casting the iree_task_allocation_header_t* buffer to a task_t

then it sounds to me like we have to change the effective type through `memcpy`, or change how we access memory.

Reply:
I looked at how these types are defined and the pointer types of the first field don't match, so they are not considered compatible under this effective type rule:
> an aggregate or union type that includes one of the aforementioned types among its members

```c
struct iree_alignas(iree_max_align_t) iree_task_t {
  // Intrusive pointer used to store tasks within iree_task_list_t and
  // iree_atomic_task_list_t singly-linked lists. This must come first in the
  // structure so that it is at the appropriate alignment.
  iree_task_t* next_task;
```

```c
typedef struct iree_task_allocation_header_t {
  // Next allocation in the linked list of allocations.
  iree_atomic_slist_intrusive_ptr_t* next;
} iree_task_allocation_header_t;
```